src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1364,80 +1364,164 @@ public override BoundNode VisitArrayCreation(BoundArrayCreation node)
+            }
+            else
+            {
+                inferredType = TypeSymbolWithAnnotations.Create(bestType);
+            }
+
+            if ((object)bestType != null && !inferredType.IsValueType)
+            {
+                // Note: so long as we have a best type, we can proceed. But we don't want to report warnings on nested nullability
+
+                // Convert elements to best type to determine element top-level nullability and to report nested nullability warnings
+                for (int i = 0; i < n; i++)
+                {
+                    var conversion = conversions[i];
+                    var element = expressions[i];
+                    var resultType = resultTypes[i];
+                    resultTypes[i] = ApplyConversion(element, element, conversion, inferredType, resultType, checkConversion: true,
+                        fromExplicitCast: false, useLegacyWarnings: false, AssignmentKind.Assignment, reportNestedWarnings: !hadNestedNullabilityMismatch, reportTopLevelWarnings: false);
+                }
+
+                // Set top-level nullability on inferred element type
+                inferredType = TypeSymbolWithAnnotations.Create(inferredType.TypeSymbol, BestTypeInferrer.GetNullableAnnotation(resultTypes));
+
                 for (int i = 0; i < n; i++)
                 {
-                    var conversion = conversionBuilder[i];
-                    var element = elementBuilder[i];
-                    var resultType = resultBuilder[i];
-                    ApplyConversion(element, element, conversion, elementType, resultType, checkConversion: true, fromExplicitCast: false, useLegacyWarnings: false, AssignmentKind.Assignment);
+                    var element = expressions[i];
+                    var resultType = resultTypes[i];
+                    // Report top-level warnings
+                    _ = ApplyConversion(element, operandOpt: null, Conversion.Identity, targetTypeWithNullability: inferredType, operandType: resultType,
+                        checkConversion: true, fromExplicitCast: false, useLegacyWarnings: false, AssignmentKind.Assignment, reportNestedWarnings: false);
                 }
             }
 
-            resultBuilder.Free();
-            elementBuilder.Free();
+            resultTypes.Free();
+            expressions.Free();
+
             _resultType = _invalidType;
+            arrayType = arrayType.WithElementType(inferredType);
             return arrayType;
         }
 
+        /// <summary>
+        /// Applies a method similar to <see cref="VisitArrayInitializer(BoundArrayCreation)"/>
+        /// The expressions returned from a lambda are not converted though, so we'll have to classify fresh conversions.
+        /// Note: even if some conversions fail, we'll proceed to infer top-level nullability. That is reasonable in common cases.
+        /// </summary>
+        internal static TypeSymbolWithAnnotations BestTypeForLambdaReturns(
```

Comment:
It looks like this method is called even for initial binding. It feels unnecessary and even risky (when nullable feature isn't enabled) to create a nullable walker and go through all this machinery.